File: apps/redux-test/src/reducers/plansReducer.ts
import { Action } from "../actions/Action";
import { initialStoreState } from "../store/initialStoreState";
import {
  updateSortedObject,
  mapObject,
  indexOfObj,
  removeKey
} from "../utils/objectUtils";
import { assertUnreachableCase } from "../utils/unreachableCase";
import { PlanMap, Plan, Timer } from "../store/IStoreState";
import { timerTimesToMs } from "../components/Timer/timerUtils";
import { resetTimers } from "../components/Timer/timerUiUtils";
import { notificationApi } from "../effects/notificationsApi";

export function plansReducer(
  _plans = initialStoreState.plans,
  action: Action
): PlanMap {
  switch (action.type) {
    case "PLANS__ADD_PLAN_ACTION":
      return { [action.newPlan.id]: action.newPlan, ..._plans };

    case "PLANS__DELETE_PLAN_ACTION":
      return removeKey(_plans, action.planId);

    case "PLANS__UPDATE_PLAN_ACTION":
      return updatePlanReducer(_plans, action.planId, action.planUpdate);

    case "PLANS__ADD_TIMER_ACTION":
      const timerPlan = _plans[action.planId];
      const newTimers = {
        [action.newTimer.id]: action.newTimer,
        ...timerPlan.timers
      };
      return updatePlanReducer(_plans, action.planId, { timers: newTimers });

    case "TIMERS__UPDATE_TIMER_ACTION":
      return updateTimerReducer(
        _plans,
        action.planId,
        action.timerId,
        action.timerUpdate
      );

    case "GLOBAL_TICK__TOCK_ACTION":
      return updateTimerTimesReducer(_plans);

    case "LOAD_STATE__GOT_LOADED_STATE":
      return action.state.plans;

    case "NAV__GO_TO_PLANS_OVERVIEW_PAGE_ACTION":
    case "NAV__GO_TO_PLAN_PAGE_ACTION":
      return _plans;
    default:
      assertUnreachableCase(action);
      return _plans;
  }
}

function updatePlanReducer(
  _plans: PlanMap,
  planId: string,
  planUpdate: Partial<Plan>
) {
  const newPlan = {
    ..._plans[planId],
    ...planUpdate
  };
  const newState = updateSortedObject(_plans, planId, newPlan);
  return newState;
}

function updateTimerReducer(
  _plans: PlanMap,
  planId: string,
  timerId: string,
  timerUpdate: Partial<Timer>
) {
  const timerPlan = _plans[planId];
  const newTimer = {
    ...timerPlan.timers[timerId],
    ...timerUpdate
  };
  const newTimers = updateSortedObject(timerPlan.timers, timerId, newTimer);
  return updatePlanReducer(_plans, planId, { timers: newTimers });
}

function updateTimerTimesReducer(_plans: PlanMap): PlanMap {
  return mapObject(_plans, (plan, _planId) => {
    switch (plan.state.status) {
      case "overview":
      case "paused":
        return plan;
      case "active":
        const activeTimerId = plan.state.activeTimer;
        const activeTimer = plan.timers[activeTimerId];
        const newTimestamp = Date.now();
        const currentTime =
          activeTimer.currentTime + (newTimestamp - plan.state.timestamp);
        const timeIsUp = currentTime >= timerTimesToMs(activeTimer.times);
        const newCurrentTime = timeIsUp
          ? timerTimesToMs(activeTimer.times)
          : currentTime;
        const newTimer = {
          ...activeTimer,
          currentTime: newCurrentTime
        };
        const newTimers = updateSortedObject(
          plan.timers,
          activeTimerId,
          newTimer
        );

        if (timeIsUp) {
          const indexOfActiveTimer = indexOfObj(newTimers, activeTimerId);
          const timerIdsArray = Object.keys(newTimers);
          const nextTimerId =
            indexOfActiveTimer != null
              ? timerIdsArray[indexOfActiveTimer + 1]
              : null;
          if (nextTimerId) {
            return {
              ...plan,
              timers: newTimers,
              state: {
                ...plan.state,
                activeTimer: nextTimerId,
                timestamp: newTimestamp
              }
            };
          } else {
            // plan is done
            const overviewPlan: Plan = {
              ...plan,
              state: { status: "overview" },
              timers: resetTimers(plan.timers)
            };
            notificationApi.notifyPlanIsDone(plan);
            return overviewPlan;
          }
        } else {
          return {
            ...plan,
            timers: newTimers,
            state: { ...plan.state, timestamp: newTimestamp }
          };
        }
    }
  });
}

File: apps/planet_plans/lib/reducers/plans_reducer.dart
import 'package:planet_plans/utils/map_utils.dart';
import 'package:planet_plans/utils/timer_ui_utils.dart';
import 'package:redux/redux.dart';

import 'package:planet_plans/models/models.dart';
import 'package:planet_plans/actions/actions.dart';

final plansReducer = combineReducers<Map<String, Plan>>([
  TypedReducer<Map<String, Plan>, PlansAddPlanAction>(_plansAddPlanReducer),
  TypedReducer<Map<String, Plan>, PlansDeletePlanAction>(
      _plansDeletePlanReducer),
  TypedReducer<Map<String, Plan>, PlansUpdatePlanAction>(
      _plansUpdatePlanReducer),
  TypedReducer<Map<String, Plan>, TimersUpdateTimerAction>(
      _timersUpdateTimerReducer),
  TypedReducer<Map<String, Plan>, GlobalTickTockAction>(_globalTickTockReducer),
  TypedReducer<Map<String, Plan>, LoadStateGotLoadedStateAction>(
      _loadStateGotLoadedStateReducer),
]);

Map<String, Plan> _plansAddPlanReducer(
    Map<String, Plan> plans, PlansAddPlanAction action) {
  return plans..addAll({action.newPlan.id: action.newPlan});
}

Map<String, Plan> _plansDeletePlanReducer(
    Map<String, Plan> plans, PlansDeletePlanAction action) {
  return plans..remove(action.planId);
}

Map<String, Plan> _plansUpdatePlanReducer(
    Map<String, Plan> plans, PlansUpdatePlanAction action) {
  return plans
    ..update(action.plan.id, (Plan plan) {
      return action.plan;
    });
}

Map<String, Plan> _timersUpdateTimerReducer(
    Map<String, Plan> plans, TimersUpdateTimerAction action) {
  return plans
    ..update(action.planId, (Plan plan) {
      final Map<String, Timer> timers = plan.timers
        ..update(action.timer.id, (Timer timer) {
          return action.timer;
        });
      return Plan(plan.state, plan.id, plan.name, timers);
    });
}

Map<String, Plan> _globalTickTockReducer(
    Map<String, Plan> plans, GlobalTickTockAction action) {
  return plans.map((String planId, Plan plan) {
    switch (plan.state.status) {
      case PlanStateStatus.overview:
      case PlanStateStatus.paused:
        return MapEntry(planId, plan);
      case PlanStateStatus.active:
        if (plan.state is PlanActiveState) {
          return _processActivePlan(plan, plan.state);
        } else {
          throw new Error();
        }
    }
  });
}

MapEntry<String, Plan> _processActivePlan(
    Plan plan, PlanActiveState planState) {
  final activeTimerId = planState.activeTimer;
  final activeTimer = plan.timers[activeTimerId];
  final newTimestamp = DateTime.now();
  final int currentTime = activeTimer.currentTime +
      (newTimestamp.difference(planState.timestamp).inMilliseconds);
  final int totalMilliseconds = activeTimer.times.toDuration().inMilliseconds;
  final timeIsUp = currentTime >= totalMilliseconds;
  final newCurrentTime = timeIsUp ? totalMilliseconds : currentTime;
  final newTimer = new Timer(activeTimer.id, activeTimer.name,
      activeTimer.times, newCurrentTime, activeTimer.category);

  final newTimers = plan.timers
    ..update(newTimer.id, (timer) {
      return newTimer;
    });

  if (timeIsUp) {
    final indexOfActiveTimer = indexInMap(newTimers, activeTimerId);
    final nextTimerExists = indexOfActiveTimer + 1 < newTimers.keys.length;
    if (nextTimerExists) {
      final nextTimerId =
          elementAtIndexInMap(newTimers, indexOfActiveTimer + 1).key;
      final newState = new PlanActiveState(nextTimerId, newTimestamp);
      return MapEntry(
          plan.id, new Plan(newState, plan.id, plan.name, newTimers));
    } else {
      // plan is done
      Plan newPlan = new Plan(
          PlanOverviewState(), plan.id, plan.name, resetTimers(plan.timers));
      // notificationApi.notifyPlanIsDone(plan);
      return MapEntry(plan.id, newPlan);
    }
  } else {
    final newState = new PlanActiveState(planState.activeTimer, newTimestamp);
    return MapEntry(plan.id, new Plan(newState, plan.id, plan.name, newTimers));
  }
}

Map<String, Plan> _loadStateGotLoadedStateReducer(
    Map<String, Plan> plans, LoadStateGotLoadedStateAction action) {
  return action.appState.plans;
}
